Make public IPs reconcile/delete async

Jont828 opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature 
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**: Implementation of an async service for public IPs as a follow up for #1610 and #1541. Thanks to @karuppiah7890 for starting this effort!

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1716 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Make public IPs reconcile/delete async
```

Jont828 commented:
/assign @CecileRobertMichon

CecileRobertMichon commented:
@Jont828 could you please rebase this one? 

Is it ready for review?

Jont828 commented:
@CecileRobertMichon Yep, I rebased and this should be ready for review!

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -146,22 +155,34 @@ func (s *ClusterScope) PublicIPSpecs() []azure.PublicIPSpec {
+				Name:           subnet.NatGateway.NatGatewayIP.Name,
+				ResourceGroup:  s.ResourceGroup(),
+				DNSName:        subnet.NatGateway.NatGatewayIP.DNSName,
+				IsIPv6:         false, // Field is implicitly false before refactor
```
Nit: can we change this comment (or remove it) so that it is more meaningful long-term (like, long after we ever remember this was ever refactored)

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -146,22 +155,34 @@ func (s *ClusterScope) PublicIPSpecs() []azure.PublicIPSpec {
+			Name:           azureBastion.PublicIP.Name,
+			ResourceGroup:  s.ResourceGroup(),
+			DNSName:        azureBastion.PublicIP.DNSName,
+			IsIPv6:         false, // Field is implicitly false before refactor
```
ditto

jackfrancis commented on azure/scope/cluster.go:
```diff
@@ -775,6 +796,9 @@ func (s *ClusterScope) FailureDomains() []string {
 		i++
 	}
+
+	sort.Strings(fds)
```
Looks like we have UT coverage for this new enforcement of a sorted string response, thanks!

jackfrancis commented on azure/services/publicips/client.go:
```diff
@@ -18,18 +18,25 @@ package publicips
 )
 
 // Client wraps go-sdk.
 type Client interface {
```
The way we implement `Client` as `AzureClient` via:

`var _ Client = &AzureClient{}`

Is weird (to me)

Why are we using this pattern in so many places?

cc @CecileRobertMichon

CecileRobertMichon commented on azure/services/publicips/client.go:
```diff
@@ -18,18 +18,25 @@ package publicips
+	Result(context.Context, azureautorest.FutureAPI, string) (result interface{}, err error)
 }
 
 // AzureClient contains the Azure go-sdk Client.
```
If possible we should try to make this type private. AFAICT it's being used in the `virtualmachines` package but now that we've refactored the `Get()` to implement the `async.Getter` interface you should be able to use the `async.Getter` interface as type for the Getter just like we're already doing for network interfaces https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2317/files#diff-bd4c935d5db79644e733f4f321bda2011252db8da4cc4bdfef5fffe3e5cd0710L55

CecileRobertMichon replied:
Are you able to make this type private (unexported) now that you removed usage in the virtualmachines package?

CecileRobertMichon replied:
My bad, you are correct. It still needs to be a public type since we're returning it in the NewClient func.

Jont828 commented:
@CecileRobertMichon @jackfrancis I removed the public IP client interface and updated it in the VM service as well. PTAL when you have some time. Thanks!

CecileRobertMichon commented:
One more comment about making the AzureClient unexported, otherwise LGTM

please squash commits

Jont828 commented:
@CecileRobertMichon I don't believe it's possible to unexport the AzureClient it as it's needed in the VM service. It's the same for network interfaces.

CecileRobertMichon reviewed (commented):
/lgtm
/approve
